cmake: Allow extending/replacing signing functionality

nordicjm opened the pull request:
Allows extended or replacing the default MCUboot signing functionality by using a cmake property which can be set by modules if alternate or modified signing functionality is required.

Fixes #54653

- [x] Implementation
- [x] Release notes update
- [x] Documentation update

nordicjm committed:
cmake: Allow extending/replacing signing functionality

Allows extended or replacing the default MCUboot signing
functionality by using a cmake property which can be set by modules
if alternate or modified signing functionality is required.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tejlmand commented on CMakeLists.txt:
```diff
@@ -1719,7 +1719,31 @@ endif()
 # Generate and use MCUboot related artifacts as needed.
 if(CONFIG_BOOTLOADER_MCUBOOT)
-  include(${CMAKE_CURRENT_LIST_DIR}/cmake/mcuboot.cmake)
+  get_property(signing_script_set
```
We have a zephyr property target for handling Zephyr build system properties, partly because it allows extracting such properties using generator expressions.

https://github.com/zephyrproject-rtos/zephyr/blob/7f0774397ed2924d21e7679ad8f452689ed7eba7/cmake/modules/kernel.cmake#L200-L205

In this case it would also remove the need for `get_property(... SET)` followed by another `get_property(...)`, cause fetching a target property returns `<var>-NOTFOUND` when not set, so test simply becomes:
```
get_target_property(signing_script zepyr_property SIGNING_SCRIPT)
if(NOT signing_script)
  set(signing_script ${CMAKE_CURRENT_LIST_DIR}/cmake/mcuboot.cmake)
  set_target_properties(zepyr_property PROPERTIES SIGNING_SCRIPT ${signing_script})
endif()

message(STATUS "Using signing script: ${signing_script}")

include(${signing_script})
```

nordicjm replied:
Neat, have updated to use that method instead.

tejlmand commented on CMakeLists.txt:
```diff
@@ -1719,7 +1719,31 @@ endif()
+    )
+
+  # If the signing_script property is not set, use the default MCUboot one included with Zephyr
+  if(${signing_script_set} EQUAL "0")
```
if keeping the `get_property(... SET)`, then the test should test against a number, but just test like `if(${signing_script_set})`, cause CMake only guarantees a boolean result, not a `0` or `1`.
```suggestion
if(${signing_script_set})
```

> If the SET option is given the variable is set to a boolean value indicating whether the property has been set.

nordicjm replied:
Have updated to just use if()

tejlmand commented on CMakeLists.txt:
```diff
@@ -1719,7 +1719,31 @@ endif()
+  endif()
+endif()
+
+# Include signing script, if set
+get_property(signing_script_set
+  GLOBAL PROPERTY
+  signing_script SET
+  )
```
I believe this `get_property(... SET)` is superfluous cause if it was unset before then you have just set the value above so no need to re-test whether it's set now.

tejlmand reviewed (changes requested):
Thanks for this, will for sure make it easier for downstream users to provide their custom signing mechanism.

Some suggestions for code simplification.

nordicjm committed:
doc: develop: west: sign: Add details on extending signing

Adds details on how to extend/replace the CMake files used for
west sign operations.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: release: 3.4: Add note on west signing extending

Adds a note that the west signing file can now be replaced.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

kartben commented on doc/develop/west/sign.rst:
```diff
@@ -89,6 +89,29 @@ Whether ``west flash`` supports this feature depends on your runner. The
+
+This will include the custom signing CMake file instead of the default zephyr
+one when projects are built with MCUboot signing support enabled. The base
+zephyr MCUboot signing file can be used as a reference for creating a new
```
```suggestion
Zephyr MCUboot signing file can be used as a reference for creating a new
```

nordicjm replied:
Fixed

kartben reviewed (changes requested):
Neat! A minor tweak + please spell Zephyr with a capital Z :) Thanks!

tejlmand reviewed (changes requested):
Let's be consistent regarding messages for including CMake files.

nordicjm commented on doc/develop/west/sign.rst:
```diff
@@ -89,6 +89,29 @@ Whether ``west flash`` supports this feature depends on your runner. The
+MCUboot enabled, signing is setup by the :file:`cmake/mcuboot.cmake` file in
+zephyr which adds extra post build commands for generating the signed images.
+The file used for signing can be replaced by adjusting the ``SIGNING_SCRIPT``
+zephyr target property - ideally done by a module using:
```
Fixed

nordicjm commented on CMakeLists.txt:
```diff
@@ -1719,7 +1719,18 @@ endif()
+endif()
+
+# Include signing script, if set
+get_target_property(signing_script zephyr_property_target SIGNING_SCRIPT)
```
> All-in-all, this updated version is much nicer than first version +1

Flow is much nicer!
